hep/proposals/0001-decorators.md

```diff
@@ -0,0 +1,656 @@
+  * Tying templates to their respective WorkflowTemplates means we know that, when used in a DAG/Steps function, the function call to another WorkflowTemplate's template to create a Task/Step should be a `TemplateRef`
+* You will be able to stub out templates from WorkflowTemplates, allowing you to use them in DAGs/Steps with full type information
+  * We intend to follow up this HEP with a "stub generation" enhancement to the CLI, where these stubs will be generated via `hera generate stubs`
+
+# Proposal
+
+This HEP proposes that Argo's DAG, Steps and Container templates be mapped into Hera as decorated functions with HeraIO classes as the inputs and outputs. We will also introduce a new script decorator under the `hera.workflows.workflow.Workflow` class which will enforce use of the script runner along with the HeraIO classes.
+
+> ⚠️ The existing context manager syntax, and the current script decorator, will continue to be supported, but they are
+> unlikely to receive any active development of new features.
+
+To allow Workflow definitions to span multiple files, we will also introduce a `TemplateSet` class offering the same decorators. This can then be used to collect templates (DAGs, scripts etc) in files `A.py` and `B.py`, while the `Workflow` object is defined in file `C.py`, which will import the `TemplateSet` from `A` and `B`, and will add all of those templates to the Workflow via a `w.add_template_set` function. This is analogous to [FastAPI's `include_router` mechanism](https://fastapi.tiangolo.com/reference/apirouter/?h=include_router#fastapi.APIRouter.include_router):
+
+```py
+from fastapi import APIRouter, FastAPI
+
+app = FastAPI()
+internal_router = APIRouter()
+users_router = APIRouter()
+
+@users_router.get("/users/")
+def read_users():
+    return [{"name": "Rick"}, {"name": "Morty"}]
+
+internal_router.include_router(users_router)
+app.include_router(internal_router)
+```
+
+# Code Examples
+
+## Basic Workflow
+
+```python
+from hera.workflows import WorkflowTemplate
+import hera.workflows.io as hio
+
+# We start by defining our Workflow Template
+wt = WorkflowTemplate(name="my-template")
+
+# Users must subclass `hio.Input` to define the template's inputs
+class MyInput(hio.Input):
+    user: str
+
+@wt.entrypoint  # Sets hello_world as the default entrypoint for the workflow template, it is an error to set entrypoint on multiple functions
+@wt.script  # Adds a new script template to the workflow template called hello_world. This will enforce use of the script runner.
+def hello_world(my_input: MyInput) -> hio.Output:  # A subclass of hio.Output must be used for the output of the function
+    output = hio.Output()
+    output.result = f"Hello Hera User: {my_input.user}!"
+    return output
+
+# `run` is a new function for users to instantiate a workflow template as a workflow
+workflow = wt.run(MyInput(name="happy-hera-user"), wait=True)
```

Comment:
What is the purpose of the wait parameter?

Where does this execute the workflow (locally or on a pre-configured Argo cluster)?

Reply:
It waits for the workflow to complete execution. In this case the workflow is running on Argo but there can be a local mode as well that is eventually planned once all of the other templates are callable.